client/commands/v2/pysa_server.py

```diff
@@ -60,6 +65,57 @@ def __init__(
         self.binary_location = binary_location
         self.server_identifier = server_identifier
 
+    def invalid_model_to_diagnostic(
+        self,
+        invalid_model: query.InvalidModel,
+    ) -> lsp.Diagnostic:
+        return lsp.Diagnostic(
+            range=lsp.Range(
+                start=lsp.Position(
+                    line=invalid_model.line - 1, character=invalid_model.column
+                ),
+                end=lsp.Position(
+                    line=invalid_model.stop_line - 1,
+                    character=invalid_model.stop_column,
+                ),
+            ),
+            message=invalid_model.full_error_message,
+            severity=lsp.DiagnosticSeverity.ERROR,
+            code=None,
+            source="Pysa",
+        )
+
+    def invalid_models_to_diagnostics(
+        self,
+        invalid_models: Sequence[query.InvalidModel],
+    ) -> Dict[Path, List[lsp.Diagnostic]]:
+        result: Dict[Path, List[lsp.Diagnostic]] = {}
+        for model in invalid_models:
+            if model.path is None:
+                self.log_and_show_message_to_client(
+                    f"{model.full_error_message}", lsp.MessageType.WARNING
+                )
+            else:
+                result.setdefault(Path(model.path), []).append(
```

Comment:
Nit: Slightly more common way of doing this is with `defaultdict` (scroll to "Using List as default_factory"):
https://www.geeksforgeeks.org/defaultdict-in-python/